### DISCUSSION TITLE
How to use the Dataloader correctly w/ Mongoose?

### DISCUSSION THREAD

**Discussion by mareksuscak** [TriPSs/nestjs-query#221] [Q&A] [answered]
Hi there. I saw the official examples and the documentation but I still do not fully understand how to leverage the built-in dataloader to address a N+1 query issue when using the Mongoose adapter. We keep seeing N+1 queries even when pulling 1st-level connections. We did not provide `batchScheduleFn` and I'm not sure when exactly it's necessary. Looking at some other examples of using the dataloader, it's rarely mentioned. I am expecting to see 2 Mongoose queries - one to retrieve the list of objects and another one to retrieve another list of objects for relationships - is this even how it's expected it to work? In reality what I see is one query to pull a list of objects and another query for each object and connection that's being accessed which seems unnecessary.

**Comment by TriPSs** [comment]
I do not know much about the Mongoose one, but willing to take a look, could you share your code?

**Reply by mareksuscak** [reply]
I'll put together a small repro based on the official Mongoose example in this repository.

**Reply by mareksuscak** [reply]
@TriPSs sorry it took me a while to get back to you. Here's the repository that illustrates the behavior. I also think I found another unexpected query - it's all mentioned in the README, you should be able to get up & running and follow the steps from the README to reproduce:

https://github.com/mareksuscak/nestjs-query-bugs

EDIT: I was basically expecting that the data loader would handle batch-expansion of related items which I managed to set up in a PR through the `loaders`, see: https://github.com/mareksuscak/nestjs-query-bugs/pull/1/files#diff-089f4f2474b64391c42b6e66aed33977e132058d92108f0a63234a7862e1f8b8R54-R80

**Reply by mareksuscak** [reply]
Hey @TriPSs, no rush, just following up on my comment above - did you have a chance to look at my minimum repro and the PR that shows a workaround? Thanks for checking!

**Reply by TriPSs** [reply]
Not yet sorry, will try to check this week!

**Reply by mareksuscak** [reply]
@TriPSs Sorry to bug you, just curious if you had a chance. If not, could you point me to where this is supposed to be handled in the source code? I'd be happy to step through that implementation to see if I can find out why it's not batch loading data.

**Reply by TriPSs** [reply]
Sorry no, you would need to checkout the `ReferenceQueryService`

**Reply by smolinari** [reply]
tldr; go to my 2nd comment below.

I'd like to chime in and granted, I'm not the super expert on this, but I'd like to share my understanding and maybe I can help. 

The dataloader library is meant to gather ids from a list of first level objects asked for in a single request (so the lifecycle of a dataloader is per request). The dataloader gathers the keys of the related (sub-)objects (so ObjectIds), in order for the child objects requested in the first level objects to be queried from the database in a single call, which is the expectation you have. I think this understanding is missing, because you are running `findOne` queries. 

So, in your example of items, a dataloader would only come into question when a query of a list of items came in ( a `find` query) and you wanted to gather all the sub-items for each item in the requested list. Your dataloader should then query the database with the keys given  (ObjectIds of the sub-items) to get objects (values) matching those keys.

Mongoose already has a dataloader like tool ready and waiting for you called [population](https://mongoosejs.com/docs/populate.html) or you could change sub-items to [sub-documents](https://mongoosejs.com/docs/subdocs.html), if you know sub-items won't grow unbounded. But since you've modelled it this way, I'm assuming sub-items can be unbounded. 

If you know the populate pattern is going to be used a lot or you wish to get recursive with your items and there is some depth involved, have a look at the [autopopulate plugin](https://www.npmjs.com/package/mongoose-autopopulate) for Mongoose. You can even control the depth of recursion (like not going deeper than X levels down).

In other words, let Mongoose do the "dataloading" part with population on one and even with many items. Population does the dataloading part for you. 

Scott  

p.s. I also saw you created a virtual for sub-item. I don't believe this will work recursively. I've never seen a virtual referring back to its own collection TBH. Usually, virtual refs are for relating between two different collections. In other words, the virtual ref would work for sure, if sub-items were in their own collection. But, not when they are in the same collection. But, maybe they would. I don't know for sure.

**Reply by smolinari** [reply]
But, now that I have cloned your repo and ran it, I see what you are getting at. :thinking: 

Let me dig deeper and get back to you. 

Scott

**Reply by smolinari** [reply]
So, I'm at a loss. All the e2e tests for Mongoose that test this functionality are commented out. Why I don't know. And unfortunately, I don't have the time to continue digging to fix this. Sorry. 

Scott

**Reply by mareksuscak** [reply]
@smolinari can you please point me at those commented tests? I couldn't find them.

**Reply by TriPSs** [reply]
@mareksuscak think he means [these](https://github.com/TriPSs/nestjs-query/blob/master/examples/mongoose/e2e/todo-item.resolver.spec.ts).

**Reply by mareksuscak** [reply]
That is what I've found but I don't see anything dataloader-related in that file unless I'm misunderstanding.

**Reply by smolinari** [reply]
@mareksuscak - As I noted in my longer post above, the Mongoose reference query service isn't using dataloaders. It's using Mongoose's populate functionality, which makes sense to use instead of dataloaders. This one place where it is used. 

https://github.com/TriPSs/nestjs-query/blob/c865ddb08ecb54c2c26fde6c61ba865ddf79b113/packages/query-mongoose/src/services/reference-query.service.ts#L155

Why your example isn't using populate is a good question. Unfortunately, I can't dig deeper at the moment. But, maybe if you can think in terms of populate, you might also find a solution for this too. I don't know. :shrug: 

Scott

**Reply by mareksuscak** [reply]
Oh okay, thanks for the additional details @smolinari! We override the generated service with a custom one that extends `MongooseQueryService` - maybe we're not doing that properly or maybe we should extend the ReferenceQueryService instead? That would be the only thing that comes to my mind. Or the wiring between the ReferenceQueryService and the MongooseQueryService isn't working as it should.

**Reply by mareksuscak** [reply, marked as answer]
@TriPSs and @smolinari: I invested some time into troubleshooting why this wasn't working and finally figured it out. This was a bug in the implementation of the Mongoose adapter and there's a similar issue in the Typegoose adapter which I am not intending to fix myself but anyone's invited to fix it based on my fix which is here: https://github.com/TriPSs/nestjs-query/pull/279

Can you please review?
